### ISSUE TITLE
Vim in WSL scrolls 3 lines when gaining focus with mouse click

### ISSUE THREAD

**Issue by francescoboc** [microsoft/terminal#5310]
# Environment

```none
Windows build number: 10.0.18363.752
Windows Terminal version (if applicable): I am using native WSL terminal

Any other software? Vim
```

# Steps to reproduce

I use Vim inside the default WSL terminal. When I click inside Vim to regain focus on the terminal, then it scrolls down 3 lines. This doesn't happen if gain focus using control tab or if I click somewhere inside Vim when I already have the terminal focused.

# Expected behavior

Vim should not scroll.

# Actual behavior

Vim scrolls down 3 lines.
Gif for clarity: https://i.redd.it/vxyk7h9bpyr41.gif

**Comment by j4james** [comment]
It may be useful to know what version of vim you're using, what plugins you have installed, and the contents of your `.vimrc` file.

**Comment by francescoboc** [comment]
Vim version: 8.0

The plugins I have installed are:
- Plugin 'itchyny/lightline.vim'
- Plugin 'scrooloose/nerdtree'
- Plugin 'tpope/vim-commentary'
- Plugin 'romainl/vim-cool'
- Plugin 'christoomey/vim-tmux-navigator'
- Plugin 'nanotech/jellybeans.vim'
- Plugin 'djoshea/vim-autoread'
- Plugin 'unblevable/quick-scope'
- Plugin 'wincent/scalpel'

But I don't think they are the problem, or anything custom I defined in my .vimrc file. I say so because I tried to comment out **everything** from my .vimrc file, leaving only the `set mouse=a` option, and the problem is still there.

**Comment by j4james** [comment]
> But I don't think they are the problem, or anything custom I defined in my .vimrc file. I say so because I tried to comment out **everything** from my .vimrc file, leaving only the `set mouse=a` option, and the problem is still there.

Yeah, you're right. I've been able to reproduce it now with just `set mouse=a`. Thanks for the additional info.

**Comment by j4james** [comment]
It looks like the problem is that we're sending a mouse up event without having sent a corresponding down event first. The reason for that is we're trying to prevent those mouse events being sent at all when focusing the window, but the code doesn't quite work the way it's intended to.

The chain of events is as follows:
1. We receive a `WM_ACTIVATE` from a click activation, and set the `CONSOLE_IGNORE_NEXT_MOUSE_INPUT` flag to indicate that we want the input handler to suppress this click.
2. We receive a `WM_LBUTTONDOWN` message, and ignore it because `CONSOLE_IGNORE_NEXT_MOUSE_INPUT` flag is set.
3. We receive a `WM_MOUSEMOVE` message, again ignoring it because `CONSOLE_IGNORE_NEXT_MOUSE_INPUT` is set. But now, because it's not a button down event, we think it's safe to reset that flag.
4. We receive a `WM_LBUTTONUP` message, but the `CONSOLE_IGNORE_NEXT_MOUSE_INPUT` flag is no longer set, so we pass that message on to the application, which triggers the bug.

So I think the fault is in the code here:
https://github.com/microsoft/terminal/blob/23f742061f5ff829540a5081c97ec997721ff780/src/interactivity/win32/windowio.cpp#L582-L587

It should probably only be resetting the flag on explicit up events, and not just anything that isn't a down event. Or at least excluding move events.

**Comment by francescoboc** [comment]
That's great! Thanks for the time you put in investigating this issue.
Just a (maybe stupid) question, on point 1, why do you do this?
> set the CONSOLE_IGNORE_NEXT_MOUSE_INPUT flag to indicate that we want the input handler to suppress this click.
Why don't you pass the click to the terminal straight away? (This is how the terminal behaves under Linux, if I remember correctly, or in any other Windows application, like Notepad for example)

**Comment by j4james** [comment]
> Why don't you pass the click to the terminal straight away?

That's a good question. I suspect that may be because the Windows console has always worked that way (i.e. apps using APIs like `ReadConsoleInput` to process mouse events). Having those apps work one way and the VT apps working another would be a bit strange.

That said, the fact that there is an inconsistency between the console and other Windows applications is also strange. I don't know. Maybe one of the MS devs will have a better explanation for the behaviour.
